File: Criticality/Code_Comments/tplfit.m
function [alpha xmin ks Loglike] = tplfit(burst,limit, x_start, vargarin)

% function fit distribution of burst to power law, and return the
% parameters which give the best power fit (with minimum KS compared with
% perfect power law
%
% alpha: exponent for power law fitting
% xmin: truncated point
% ks: value for Kolmogorov–Smirnov test
%
% parameters:
% burst: could be either avalanche sizes or durations (a vector)
% limit set up the range from which we could get the xmin
% x_start fix the xmin as x_start

KS = []; alpha = [];  Loglike = [];

switch nargin
    case 2
        range = 1:limit;
    case 3
        range = x_start;
end


for x0 = range
    
    % [alpha xmin]=plfit(burst);
    X = burst;
    xmax = max(burst);
    n = length(burst(x0<=burst & burst<=xmax)) ; % number of avalanches
    s = unique(X(x0<=X & X<=xmax)); 
    smin = min(s); % minimum avalanche
    smax = max(s); % maximum avalanche
    
    LL = @(x) x*sum( log( burst(x0<=burst & burst<=xmax) ) ) - n*log( 1/sum(s.^-x ) ) ; % Maximum likelihood
    [a,fval] = fminsearch(LL , 2.3 ); % start search from 2.3    
    Loglike = [Loglike , -fval]; % value of the minimum likelihood
    
    
    % X = burst;
    %
    % for i = 1:1000
    %
    %     X = Shuffle(X);
    %     I = randi(length(X),[1 length(X)]);
    %     AV = X(I);
    %     [sth xmin]=plfit(AV);
    % %     xmax = max(AV);
    %     n = length(AV(xmin<=AV & AV<=xmax)) ;
    %     s = unique(AV(xmin<=AV & AV<=xmax));
    %     smin = min(s);      smax = max(s);
    %     LL = @(x) x*sum( log( AV(xmin<=AV & AV<=xmax) ) ) - n*log( 1/sum(s.^-x ) ) ;
    % %     options = optimset('PlotFcns',@optimplotfval);
    %     [a2,fval] = fminsearch(LL , 2.8);
    %     alpha = [alpha , a2];
    %     clear a2 I
    % end
    %
    % SD = std(alpha);   
    %% ########################## Hypothesis Testing ###############################
    % alpha = a;
    % [sth xmin] = plfit(burst);
    xmax = max(burst);    
    N = length(burst);
     k = 0;    
    z = burst;
    z = z(z>=x0);            n    = length(z);
    cdf = cumsum(hist(z,x0:xmax)./n); % cdf
    
    s = unique(X(x0<=X & X<=xmax));
     smin = min(s);      smax = max(s);
    A = 1/ sum(s.^-a ); % constant for perfect power law
    alpha = [alpha , a];
    fit{x0} = cumsum (A*(x0:xmax).^-a ); % perfect power law PDF
    KS = [KS , max(abs(cdf - fit{x0})) ] ; % calculate Kolmogorov–Smirnov

    
end

switch nargin
    case 2
        xmin = find(KS==min(KS)) ;  alpha = alpha( xmin );  Loglike = Loglike(xmin);
    case 3
        xmin = x_start;
end


ks = min(KS);

 

File: Criticality/tplfit.py
import numpy as np
from copy import deepcopy as cdc
import scipy.optimize
from sahara_work import Criticality as cr
def tplfit(burst,limit):
	KS = []
	alpha = []
	Loglike = []
	alpha = []

	for x0 in np.arange(1, limit+1):

		X = cdc(burst)
		xmax = np.max(burst)
		idx = np.where(np.logical_and(x0<=X, X<=xmax))[0]
		n = np.size(burst[idx])
		s = np.unique(X[idx]) 
		smin = np.min(s) # minimum avalanche
		smax = np.max(s) # maximum avalanche

		LL = lambda x: x*np.sum(np.log(burst[idx])) - n*np.log(1/np.sum(np.power(s,-x)))
		a = scipy.optimize.fmin(func=LL, x0=2.3, disp = False) # start search from 2.3
		fval = LL(a) #value of the minimum likelihood
		Loglike.append(-fval)

		N = np.size(burst)
		k = 0

		z = cdc(burst)
		z = z[z>=x0]
		n = np.size(z)
		cdf = np.cumsum(np.histogram(z,bins = np.arange(x0,xmax+2))[0]/n) # cdf
		A = 1/(np.sum(np.power(s, -a)))
		alpha.append(a)
		fit = np.cumsum(A*np.power(np.arange(x0,xmax+1), -a))
		KS.append(np.max(np.abs(cdf-fit)))

	xmin = int(np.where(KS==np.min(KS))[0])
	alpha = alpha[xmin]
	Loglike = Loglike[xmin];
	ks = np.min(KS)
	xmin = xmin + 1

	return alpha, xmin, ks, Loglike
